src/main/scala/devices/debug/Debug.scala

```diff
@@ -1192,40 +1194,38 @@ class TLDebugModuleInner(device: Device, getNComponents: () => Int, beatBytes: I
     // by harts executing code.
     //--------------------------------------------------------------
 
-    for (component <- 0 until nComponents) {
-      when (~io.dmactive || ~dmAuthenticated) {
-        haltedBitRegs(component) := false.B
-        resumeReqRegs(component) := false.B
+    when (~io.dmactive || ~dmAuthenticated) {
+        haltedBitRegs := 0.U
+        resumeReqRegs := 0.U
       }.otherwise {
-        // Hart Halt Notification Logic
-        when (hartIsInResetSync(component)) {
-          haltedBitRegs(component) := false.B
-          resumeReqRegs(component) := false.B
-        }.elsewhen (hartHaltedWrEn) {
-          when (hartSelFuncs.hartIdToHartSel(hartHaltedId) === component.U) {
-            haltedBitRegs(component) := true.B
-          }
+
+        haltedBitRegs := haltedBitRegs & ~(hartIsInResetSync.asUInt)
+        resumeReqRegs := resumeReqRegs & ~(hartIsInResetSync.asUInt)
+
+        val hartHaltedIdIndex   = UIntToOH(hartSelFuncs.hartIdToHartSel(hartHaltedId))
+        val hartResumingIdIndex = UIntToOH(hartSelFuncs.hartIdToHartSel(hartResumingId))
+        val hartselIndex        = UIntToOH(io.innerCtrl.bits.hartsel)
+        when (hartHaltedWrEn) {
+          haltedBitRegs := haltedBitRegs | hartHaltedIdIndex
         }.elsewhen (hartResumingWrEn) {
-          when (hartSelFuncs.hartIdToHartSel(hartResumingId) === component.U) {
-            haltedBitRegs(component) := false.B
-          }
+          haltedBitRegs := haltedBitRegs & ~(hartResumingIdIndex)
         }
```

Comment:
This will fail if one hart's hartIsInResetSync is 1 in the same cycle that another hart is doing hartHaltedWrEn.  I suggest combining everything into one expression, like this:

```
        when (hartHaltedWrEn) {
          haltedBitRegs := (haltedBitRegs | hartHaltedIdIndex) & ~(hartIsInResetSync.asUInt)
        }.elsewhen (hartResumingWrEn) {
          haltedBitRegs := haltedBitRegs & ~(hartResumingIdIndex) & ~(hartIsInResetSync.asUInt)
        }.otherwise {
          haltedBitRegs := haltedBitRegs & ~(hartIsInResetSync.asUInt)
        }
```

Reply:
I tried this out, but the asm-tests run indefinitely in this case.

Reply:
Correction.
It was when I tried applying the same transformation to resumeReqRegs that the asm-tests run indefintely.
For now pulling only haltedBitRegs